logging/logging.go

```diff
@@ -298,13 +298,75 @@ func detectGAEResource() *mrpb.MonitoredResource {
 	}
 }
 
+func isCloudRun() bool {
+	_, config := os.LookupEnv("K_CONFIGURATION")
+	return config
+}
+
+func detectCloudRunResource() *mrpb.MonitoredResource {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil
+	}
+	return &mrpb.MonitoredResource{
+		Type: "cloud_run_revision",
+		Labels: map[string]string{
+			"project_id":         projectID,
+			"location":           regionFromZone(zone),
+			"service_name":       os.Getenv("K_SERVICE"),
+			"revision_name":      os.Getenv("K_REVISION"),
+			"configuration_name": os.Getenv("K_CONFIGURATION"),
+		},
+	}
+}
+
+func isCloudFunction() bool {
+	_, name := os.LookupEnv("FUNCTION_NAME")
+	_, target := os.LookupEnv("FUNCTION_TARGET")
+	if name || target {
+		return true
+	}
+	return false
+}
+
+func detectCloudFunction() *mrpb.MonitoredResource {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil
+	}
+	// Newer functions runtimes store name in K_SERVICE
+	functionName, exists := os.LookupEnv("K_SERVICE")
+	if !exists {
+		functionName, _ = os.LookupEnv("FUNCTION_NAME")
+	}
+	return &mrpb.MonitoredResource{
+		Type: "cloud_function",
+		Labels: map[string]string{
+			"project_id":    projectID,
+			"region":        regionFromZone(zone),
+			"function_name": functionName,
+		},
+	}
+}
+
 func detectResource() *mrpb.MonitoredResource {
 	detectedResource.once.Do(func() {
 		switch {
-		// GAE needs to come first, as metadata.OnGCE() is actually true on GAE
-		// Second Gen runtimes.
+		// GAE, CloudRun need to come first, as metadata.OnGCE() returns true on these runtimes.
 		case os.Getenv("GAE_ENV") == "standard":
 			detectedResource.pb = detectGAEResource()
```

Comment:
what happens in AppEngine Flex?

Reply:
This was already implemented in place before this PR. I believe it doesn't detect flex, and it incorrectly labels flex logs as GCE. =[

I plan to fix this in the future, once I have environment tests configured for app engine.

Reply:
Let's open a tracking bug to fix AppEngine Flex detection so we don't forget it

Reply:
Added here: https://github.com/googleapis/google-cloud-go/issues/3919